File: coding/src/com/algos/li05_binarysearch/revised/L774_MinimizeMaxDistancetoGasStation.java
package com.algos.li05_binarysearch.revised;

public class L774_MinimizeMaxDistancetoGasStation {
        
    public static void main(String[] args) {
        L774_MinimizeMaxDistancetoGasStation obj = new L774_MinimizeMaxDistancetoGasStation();
        int[] stations = {1,2,3,4,5,6,7,8,9,10};
        int k = 9;
        System.out.println(obj.minmaxGasDist(stations, k));
    }
    
    // Problem: https://leetcode.ca/2018-01-12-774-Minimize-Max-Distance-to-Gas-Station/
    // Idea: We need to apply BS on the range (0, max distance between two stations) and find the minimum distance such that we can add k stations.
    // Time: O(nlogn)
    public double minmaxGasDist(int[] stations, int k) {
        int n = stations.length;
        double low = 0, high = 0;
        for(int i=0; i<n-1; i++){   // max diff between two stations
            high = Math.max(high, stations[i+1] - stations[i]);
        }
        
        double ans = -1;
        while(high - low > 1e-6){
            double mid = (low + high)/2.0;
            int cnt = noOfStationsRequired(stations, mid);
            if(cnt > k){
                low = mid;
            }else{
                ans = mid;
                high = mid;
            }
        }

        return ans;
    }
    
    private int noOfStationsRequired(int[] stations, double dist) {
        int cnt = 0;
        for(int i=0; i<stations.length-1; i++){
            cnt += Math.ceil((stations[i+1] - stations[i])/dist) - 1;
        }
        return cnt;
    }
}

File: dsa/patterns/05_binarysearch/python/l774_minimize_max_distance_to_gas_station.py
# LeetCode 774 - Minimize Max Distance to Gas Station
#
# Idea:
# Binary search on the answer (maximum allowed distance).
# For a guessed distance 'd', compute how many extra stations
# are required so that no segment exceeds 'd'.
#
# Time Complexity: O(n * log(range / precision))
# Space Complexity: O(1)

import math


def stations_required(stations: list[int], max_dist: float) -> int:
    required = 0

    for i in range(len(stations) - 1):
        gap = stations[i + 1] - stations[i]
        required += math.ceil(gap / max_dist) - 1   # [1, 2] 0.5 dist, we can 1

    return required


def minmax_gas_dist(stations: list[int], k: int) -> float:
    left = 0.0

    right = 0.0
    for i in range(len(stations) - 1):
        right = max(right, stations[i + 1] - stations[i])  # Max dist of adj

    ans = 0.0
    eps = 1e-6

    while right - left > eps:
        mid = (left + right) / 2.0

        if stations_required(stations, mid) <= k:  # k or fewer stations
            ans = mid
            right = mid  # minimize dist
        else:
            left = mid

    return ans


# Driver code
if __name__ == "__main__":
    stations = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    k = 9
    print(minmax_gas_dist(stations, k))  # ~0.5
